File: leetcode/C++/sliding-window-maximum.cpp
class Solution {
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        deque<int> dq;
        vector<int> sol;
        for(int i=0; i<nums.size(); ++i){
            while(!dq.empty() && nums[dq.back()]<=nums[i]) dq.pop_back();
            if(!dq.empty() && i-dq.front()+1>k) dq.pop_front();
            dq.push_back(i);
            if(i>=k-1) sol.push_back(nums[dq.front()]);
        }
        return sol;
    }
};

File: leetcode/Python/sliding-window-maximum.py
class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        dq, sol = collections.deque(), []
        for i, v in enumerate(nums):
            while dq and nums[dq[-1]]<=v:
                dq.pop()
            if dq and i-dq[0]+1>k:
                dq.popleft()
            dq.append(i)
            if i>=k-1:
                sol.append(nums[dq[0]])
        return sol